spec/support/shared_examples/block_matcher.rb

```diff
@@ -0,0 +1,102 @@
+RSpec.shared_examples "an RSpec block-only matcher" do
+  # Note: do not use `matcher` in 2 expectation expressions in a single
+  # example here. In some cases (such as `change { }.to(2)`), it will fail
+  # because using it a second time will apply `x += 2` twice, changing
+  # the value to 4.
+
+  matcher :always_passes do
+    supports_block_expectations
+    match do |actual|
+      actual.call
+      true
+    end
+  end
+
+  matcher :always_fails do
+    supports_block_expectations
+    match do |actual|
+      actual.call
+      false
+    end
+  end
+
+  def valid_expectation
+    expect { |block| valid_block(&block) }
+  end
+
+  def invalid_expectation
+    expect { |block| invalid_block(&block) }
+  end
+
+  include_examples "an RSpec matcher"
+
+  it 'preserves the symmetric property of `==`' do
+    expect(matcher).to eq(matcher)
+    # TODO: the following expectations fail due to
+    #
+    # 1) output.to_stderr matcher behaves like an RSpec block-only matcher preserves the symmetric property of `==`
+    #    Failure/Error: raise "Warnings were generated: #{output}" if has_output?
+    #    RuntimeError:
+    #      Warnings were generated: foo
+    #
+    # expect(matcher).not_to eq(valid_block)
+    # expect(valid_block).not_to eq(matcher)
+    #
+    # Does it make sense for block matchers?
+  end
+
+  # TODO: Two options here.
+  #
+  # The examples below fail in yield_spec.rb, total 8 failures.
+  #
+  # it 'matches a valid block when using #=== so it can be composed' do
+  #   expect(matcher).to be === -> { valid_block }
+  # end
+  #
+  # it 'does not match an invalid block when using #=== so it can be composed' do
+  #   expect(matcher).not_to be === -> { invalid_block }
+  # end
+  #
+  #
+  # The examples below (with `block` passed) pass for yield_spec.rb, but
+  # fail in other block matcher specs (10 failures total).
+  #
+  # it 'matches a valid block when using #=== so it can be composed' do
+  #   expect(matcher).to be === ->(block) { valid_block(&block) }
+  # end
+  #
+  # it 'does not match an invalid block when using #=== so it can be composed' do
+  #   expect(matcher).not_to be === ->(block) { invalid_block(&block) }
+  # end
+  #
+  # I see two options:
+  # 1. Introduce another conditional flag, and run the former two examples
+  #    for all specs except yield_spec.rb. Also pull those two latter to
+  #    yield_spec.rb itself.
+  # 2. Remove this. I'm not too confident with the code base to tell if it's
+  #    absolutely necessary to check this, and why `===` is vital for
+  #    composing matchers, and if it's applicable to block matchers.
```

Comment:
Its applicable to all matchers, its the protocol for them.

Reply:
Exactly the same problem as below it seems. `proc` fixed it.